apps/data-extraction/src/extractors/submarine-ranks.extractor.ts

```diff
@@ -0,0 +1,29 @@
+import { AbstractExtractor } from '../abstract-extractor';
+import { map, tap } from 'rxjs/operators';
+import { XivapiEndpoint } from '@xivapi/angular-client';
+
+export class SubmarineRanksExtractor extends AbstractExtractor {
+
+  protected doExtract(): any {
+    const ranks = {};
+
+    this.getAllPages(this.getResourceEndpointWithQuery(XivapiEndpoint.SubmarineRank, {
+      columns: 'ID,Capacity,ExpToNext,SurveillanceBonus,RetrievalBonus,SpeedBonus,RangeBonus,FavorBonus',
+    })).pipe(
+      map(((page) => page.Results)),
+      tap((rankResults)  => {
+        rankResults.forEach((result) => {
+          ranks[result.ID] = result;
```

Comment:
Whenever it's possible, prefer a custom structure so if tomorrow things are renamed for whatever reason, only this part of the code will have to change. This also helps with naming convention as TC uses lowercase for property names while XIVAPI uses UpperCamelCase.

Reply:
Np. Isn't there something that can be used for checking the struct so the extractor throws an error if the response doesn't contain the columns? Like an endpoint in XIVAPI that returns the body spec or just the list of the fields for a given index (eg: SubmarineRanks)
That would avoid the trouble of checking if there is struct changes manually every update.

Reply:
I'm the maintainer of XIVAPI so I'll be aware of changes since I'm the one making them 😉 to get the fields for a given sheet, just get one of the items, like https://xivapi.com/SubmarineRank/1, there's no optional fields, so every single field can be found on every single entry 👍